pyro/contrib/funsor/infer/traceenum_elbo.py

```diff
@@ -148,19 +148,33 @@ def differentiable_loss(self, model, guide, *args, **kwargs):
             costs = contracted_costs + uncontracted_factors  # model costs: logp
             costs += [-f for f in guide_terms["log_factors"]]  # guide costs: -logq
 
-            # finally, integrate out guide variables in the elbo and all plates
             plate_vars = guide_terms["plate_vars"] | model_terms["plate_vars"]
-            elbo = to_funsor(0, output=funsor.Real)
+            # compute the marginal logq in the guide corresponding to each cost term
+            targets = dict()
             for cost in costs:
-                # compute the marginal logq in the guide corresponding to this cost term
-                log_prob = funsor.sum_product.sum_product(
+                input_vars = frozenset(cost.inputs.keys())
+                if input_vars not in targets:
+                    targets[input_vars] = funsor.Tensor(
+                        funsor.ops.new_zeros(
+                            funsor.tensor.get_default_prototype(), ()
+                        ).expand(tuple(v.size for v in cost.inputs.values())),
+                        cost.inputs,
+                        cost.dtype,
+                    )
+            with AdjointTape() as tape:
+                logzq = funsor.sum_product.sum_product(
                     funsor.ops.logaddexp, funsor.ops.add,
-                    guide_terms["log_measures"],
+                    guide_terms["log_measures"] + list(targets.values()),
                     plates=plate_vars,
-                    eliminate=(plate_vars | guide_terms["measure_vars"]) - frozenset(cost.inputs)
+                    eliminate=(plate_vars | guide_terms["measure_vars"])
                 )
-                # compute the expected cost term E_q[logp] or E_q[-logq] using the marginal logq for q
-                elbo_term = funsor.Integrate(log_prob, cost, guide_terms["measure_vars"] & frozenset(cost.inputs))
+            marginals = tape.adjoint(funsor.ops.logaddexp, funsor.ops.add, logzq, tuple(targets.values()))
+            # finally, integrate out guide variables in the elbo and all plates
+            elbo = to_funsor(0, output=funsor.Real)
+            for cost in costs:
+                target = targets[frozenset(cost.inputs.keys())]
+                log_prob = marginals[target] - logzq
```

Comment:
Hmm... I think you can divide by the plate-local normalizing constants here rather than the global `logzq`
```py
logzq_local = marginals[target].reduce(funsor.ops.logaddexp, cost.input_vars - plate_vars)
log_prob = marginals[target] - logzq
```
which should lead to lower gradient variance.  I believe this is what the main implementation of `TraceEnum_ELBO` does. More generally, this would benefit from eager normalization within the backward pass of `adjoint`.

Reply:
All three of these work (same loss and grad):

```py
log_prob = marginals[target] 
log_prob = marginals[target] - logzq
log_prob = marginals[target] - logzq_local
```

even when compared against the main `TraceEnum_ELBO`:

```py
from pyro.infer import TraceEnum_ELBO
...
    pyro_elbo = TraceEnum_ELBO(max_plate_nesting=2)
    pyro_loss = pyro_elbo.differentiable_loss(model_plate, guide_plate)
    _check_loss_and_grads(actual_loss, pyro_loss)
```

I'm still trying to understand why. Is it because in these examples guide is always normalized?

Reply:
Because guides are always normalized and all of the integration in these tests is exact (i.e. there is no sampling).